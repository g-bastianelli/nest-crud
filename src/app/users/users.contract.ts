import { initContract } from '@ts-rest/core';
import { z } from 'zod';
import { paginationResponseSchema } from '../utils/pagination.utils';

const userSchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  email: z.string().email(),
  claimsTotal: z.number(),
});

const c = initContract();

const usersContract = c.router({
  createUser: {
    method: 'POST',
    path: '/users',
    responses: {
      201: userSchema,
    },
    body: z.object({
      name: z.string(),
      email: z.string().email(),
    }),
    summary: 'Create a user',
  },
  getUser: {
    method: 'GET',
    path: `/users/:id`,
    pathParams: z.object({
      id: z.string().uuid(),
    }),
    responses: {
      200: userSchema.nullable(),
    },
    summary: 'Get a user by id',
  },
  listUsers: {
    method: 'GET',
    path: '/users',
    query: z.object({
      page: z.coerce.number().min(1).default(1),
      pageSize: z.coerce.number().min(1).max(100).default(10),
    }),
    responses: {
      200: z.object({
        users: z.array(userSchema),
        pagination: paginationResponseSchema,
      }),
    },
    summary: 'List all users',
  },
});

export { usersContract };
