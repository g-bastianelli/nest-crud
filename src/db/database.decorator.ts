import { Inject } from '@nestjs/common';
import { TEST_CRUD_DB } from './database.consts';

const InjectDatabase = () => Inject(TEST_CRUD_DB);

export { InjectDatabase };
