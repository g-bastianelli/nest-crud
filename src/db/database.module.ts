import { Module } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { drizzle } from 'drizzle-orm/node-postgres';
import * as schema from './database.schema';
import { DatabaseService } from './database.service';
import { TEST_CRUD_DB } from './database.consts';

@Module({
  providers: [
    {
      provide: TEST_CRUD_DB,
      inject: [ConfigService],
      useFactory: async (config: ConfigService) => {
        return drizzle(config.get('DATABASE_URL'), { schema });
      },
    },
    DatabaseService,
  ],
  exports: [TEST_CRUD_DB],
})
class DatabaseModule {}

export { DatabaseModule };
