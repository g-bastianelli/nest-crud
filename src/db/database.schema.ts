import { integer, pgTable, uuid, varchar } from 'drizzle-orm/pg-core';

export const users = pgTable('users', {
  id: uuid().defaultRandom().primaryKey(),
  name: varchar({ length: 255 }).notNull(),
  email: varchar({ length: 255 }).notNull().unique(),
});

export const claims = pgTable('claims', {
  id: uuid('id').defaultRandom().primaryKey().notNull(),
  userId: uuid()
    .references(() => users.id)
    .notNull(),
  title: varchar({ length: 255 }).notNull(),
  description: varchar({ length: 255 }).notNull(),
  value: integer().notNull(),
});
