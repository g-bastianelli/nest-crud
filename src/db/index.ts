export * from './database.consts';
export * from './database.decorator';
export * from './database.module';
export * from './database.types';

import * as schema from './database.schema';

export { schema };
