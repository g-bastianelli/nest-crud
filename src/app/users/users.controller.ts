import { Controller, Req } from '@nestjs/common';
import { TsRestException, tsRestHandler, TsRestHandler } from '@ts-rest/nest';
import { usersContract } from './users.contract';
import { UsersService } from './users.service';
import { buildPaginationResponse } from '../utils/pagination.utils';

@Controller()
class UsersController {
  constructor(private readonly usersService: UsersService) {}

  @TsRestHandler(usersContract.getUser)
  getUser() {
    return tsRestHandler(usersContract.getUser, async ({ params: { id } }) => {
      const user = await this.usersService.getUserByIdWithClaimsTotal(id);
      if (!user) {
        throw new TsRestException(usersContract.createUser, {
          status: 404,
          body: {
            code: 'USER_NOT_FOUND',
            message: 'User not found',
          },
        });
      }
      return {
        status: 200,
        body: user,
      };
    });
  }

  @TsRestHandler(usersContract.listUsers)
  listUsers(@Req() request: Request) {
    return tsRestHandler(
      usersContract.listUsers,
      async ({ query: { page, pageSize } }) => {
        const { users, total } = await this.usersService.listUsers(
          page,
          pageSize,
        );
        return {
          status: 200,
          body: {
            users,
            pagination: buildPaginationResponse(request, {
              page,
              pageSize,
              total,
            }),
          },
        };
      },
    );
  }
}

export { UsersController };
