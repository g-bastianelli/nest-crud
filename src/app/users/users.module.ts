import { Module } from '@nestjs/common';
import { UsersController } from './users.controller';
import { DatabaseModule } from '@db';
import { UsersService } from './users.service';

@Module({
  imports: [DatabaseModule],
  controllers: [UsersController],
  providers: [UsersService],
})
class UsersModule {}

export { UsersModule };
