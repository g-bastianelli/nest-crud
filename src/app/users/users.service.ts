import { Injectable } from '@nestjs/common';
import { DatabaseType, InjectDatabase, schema } from '../../db';
import { eq, sql } from 'drizzle-orm';

@Injectable()
export class UsersService {
  constructor(@InjectDatabase() private readonly db: DatabaseType) {}

  async getUserByIdWithClaimsTotal(userId: string) {
    const [user] = await this.getUserWithClaimsTotalQuery(userId);
    return user ?? null;
  }

  async listUsers(page: number, pageSize: number) {
    const [{ count }] = await this.countUsersQuery();

    const users = await this.listUsersQuery(page, pageSize);

    return { users, total: count };
  }

  private countUsersQuery() {
    return this.db
      .select({
        count: sql<number>`count(*)`,
      })
      .from(schema.users);
  }

  private getUserWithClaimsTotalQuery(userId: string) {
    return this.db
      .select({
        id: schema.users.id,
        name: schema.users.name,
        email: schema.users.email,
        claimsTotal: sql<number>`sum(${schema.claims.value})`,
      })
      .from(schema.users)
      .innerJoin(schema.claims, eq(schema.claims.userId, schema.users.id))
      .where(eq(schema.users.id, userId))
      .groupBy(schema.users.id)
      .limit(1);
  }

  private listUsersQuery(page: number, pageSize: number) {
    return this.db.query.users.findMany({
      limit: pageSize,
      offset: (page - 1) * pageSize,
      orderBy: (user, { asc }) => asc(user.name),
    });
  }
}
