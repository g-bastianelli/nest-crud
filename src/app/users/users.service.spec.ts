import { describe } from 'node:test';
import { UsersService } from './users.service';
import { drizzle } from 'drizzle-orm/node-postgres';
import { schema } from '../../db';

describe('UsersService', () => {
  const usersService = new UsersService(drizzle('mock', { schema }));

  it('listUsersQuery should generate the right query', function () {
    expect(usersService['listUsersQuery'](22, 1000).toSQL()).toMatchSnapshot();
  });

  it('getUserWithClaimsTotalQuery should generate the right query ', function () {
    expect(
      usersService['getUserWithClaimsTotalQuery']('1').toSQL(),
    ).toMatchSnapshot();
  });

  it('countUsersQuery should generate the right query', function () {
    expect(usersService['countUsersQuery']().toSQL()).toMatchSnapshot();
  });
});
