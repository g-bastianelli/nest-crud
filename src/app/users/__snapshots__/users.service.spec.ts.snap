// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`countUsersQuery should generate the right query 1`] = `
{
  "params": [],
  "sql": "select count(*) from "users"",
}
`;

exports[`getUserWithClaimsTotalQuery should generate the right query  1`] = `
{
  "params": [
    "1",
    1,
  ],
  "sql": "select "users"."id", "users"."name", "users"."email", sum("claims"."value") from "users" inner join "claims" on "claims"."userId" = "users"."id" where "users"."id" = $1 group by "users"."id" limit $2",
}
`;

exports[`listUsersQuery should generate the right query 1`] = `
{
  "params": [
    1000,
    21000,
  ],
  "sql": "select "id", "name", "email" from "users" order by "users"."name" asc limit $1 offset $2",
  "typings": [
    "none",
    "none",
  ],
}
`;
